[MINOR] Add documentation for using multi-column table keys and for n… (#1761)
apache/hudi: docs/_docs/2_2_writing_data.md

## Datasource Writer

The `hudi-spark` module offers the DataSource API to write (and read) a Spark DataFrame into a Hudi table. There are a number of options available:

**`HoodieWriteConfig`**:

**TABLE_NAME** (Required)<br>

**`DataSourceWriteOptions`**:

**RECORDKEY_FIELD_OPT_KEY** (Required): Primary key field(s). Nested fields can be specified using the dot notation eg: `a.b.c`. When using multiple columns as primary key use comma seperated notaion, eg: `"col1,col2,col3,etc"`. Single or multiple columns as primary key specified by `KEYGENERATOR_CLASS_OPT_KEY` property.<br>
Default value: `"uuid"`<br>

> **PARTITIONPATH_FIELD_OPT_KEY** (Required): Columns to be used for partitioning the table. To prevent partitioning, provide empty string as value eg: `""`. Specify paritioning/no partitioning using `HIVE_PARTITION_EXTRACTOR_CLASS_OPT_KEY`<br>

Reviewer: `HIVE_PARTITION_EXTRACTOR_CLASS_OPT_KEY` should be changed to `KEYGENERATOR_CLASS_OPT_KEY`? and also when specify no partitioning tables, would use `NonpartitionedKeyGenerator`

Author: Yap, seems both are necessary:
```suggestion
**PARTITIONPATH_FIELD_OPT_KEY** (Required): Columns to be used for partitioning the table. To prevent partitioning, provide empty string as value eg: `""`. Specify partitioning/no partitioning using both `KEYGENERATOR_CLASS_OPT_KEY` and `HIVE_PARTITION_EXTRACTOR_CLASS_OPT_KEY`<br>
```

Reviewer: > Yap, seems both are necessary:

The `HIVE_PARTITION_EXTRACTOR_CLASS_OPT_KEY` is used to sync to hive, if you are meaning sync to hive during the write, it makes sense, or it makes no sense.

Author: I see under the "syncing to hive" section it states "Both tools above support syncing of the table’s latest schema to Hive metastore" which implies that syncing table metadata to hive is not mandatory but optional. Is this actually true? It seems that using a metastore is actually mandatory...

Reviewer: One would only write data into hudi but not syncing to hive, so the `HIVE_PARTITION_EXTRACTOR_CLASS_OPT_KEY ` is only used when syncing to hive metastore.

Author: Will add clarification that HIVE_PARTITION_EXTRACTOR_CLASS_OPT_KEY is only necessary when using hive.

> **PARTITIONPATH_FIELD_OPT_KEY** (Required): Columns to be used for partitioning the table. To prevent partitioning, provide empty string as value eg: `""`. Specify partitioning/no partitioning using `KEYGENERATOR_CLASS_OPT_KEY` and if using hive `HIVE_PARTITION_EXTRACTOR_CLASS_OPT_KEY`<br>

Reviewer: hi, again the `HIVE_PARTITION_EXTRACTOR_CLASS_OPT_KEY` is not needed when not syncing to hive.

Author: This notes that HIVE_PARTITION_EXTRACTOR_CLASS_OPT_KEY only needs to be specified if using hive. Do you have a suggestion on how to phrase it differently?

Reviewer: how about 
Specify partitioning/no partitioning using `KEYGENERATOR_CLASS_OPT_KEY` and using `HIVE_PARTITION_EXTRACTOR_CLASS_OPT_KEY` when syncing to hive ?

Author: Building on that, how about:?
Specify partitioning/no partitioning using `KEYGENERATOR_CLASS_OPT_KEY`. If synchronizing to hive, also specify using `HIVE_PARTITION_EXTRACTOR_CLASS_OPT_KEY`.

Reviewer: > Building on that, how about:?
> Specify partitioning/no partitioning using `KEYGENERATOR_CLASS_OPT_KEY`. If synchronizing to hive, also specify using `HIVE_PARTITION_EXTRACTOR_CLASS_OPT_KEY`.

resonable.

Default value: `"partitionpath"`<br>

**PRECOMBINE_FIELD_OPT_KEY** (Required): When two records have the same key value, the record with the largest value from the field specified here will be choosen.<br>
Default value: `"ts"`<br>

> **OPERATION_OPT_KEY**: The [write operations](#write-operations) to use. Note: this cannot change across writes.<br>

Reviewer: hi, why this cannot change across writes? would you please clarify? IIRU, It would be changed across writes.

Author: Good catch, that was meant to go in the description of the next parameter, TABLE_TYPE_OPT_KEY.

The text can be made more clear by using the following language instead
Instead of "Note: this cannot change across writes."
Use "Note: After the initial creation of a table, this value must stay consistent when writing to the table using the Spark SaveMode.Append mode."

Also, I believe that the key and partition columns, cannot be changed after the table is first created as well, so this note snippet can be added to those fields as well. Is this correct?

Author: Btw, there seems to be a mismatch in the data deletion documentation as well. To delete records it seems the only thing necessary is to set OPERATION_OPT_KEY to DELETE_OPERATION_OPT_VAL and HIVE_PARTITION_EXTRACTOR_CLASS_OPT_KEY to `classOf[GlobalDeleteKeyGenerator].getCanonicalName`, however https://hudi.apache.org/docs/writing_data.html#deletes specifies a hard delete procedure that is different from this one.

Reviewer: in fact, there are some ways to delete records in hudi, 1. we would use "org.apache.hudi.EmptyHoodieRecordPayload" to delete a batch of records, 2. we could directly use `delete` api, 3. we could we `_hoodie_is_delete` flag to delete record. Also the `HIVE_PARTITION_EXTRACTOR_CLASS_OPT_KEY ` has nothing to do with the writing path IIRC, it is only related to hive sync.

Reviewer: > Good catch, that was meant to go in the description of the next parameter, TABLE_TYPE_OPT_KEY.
> 
> The text can be made more clear by using the following language instead
> Instead of "Note: this cannot change across writes."
> Use "Note: After the initial creation of a table, this value must stay consistent when writing to (updating) the table using the Spark SaveMode.Append mode."
> 
> Also, I believe that the key and partition columns, cannot be changed after the table is first created as well, so this note snippet can be added to those fields as well. Is this correct?

Yes, `TABLE_TYPE_OPT_KEY` is not meant to be changed in next write, and `OPERATION_OPT_KEY` would be changed during different writes, and maybe I misunderstand what you meant.

Author: In regards to deletion: I believe method 2 that you mentioned with delete api, is the same thing as as setting  OPERATION_OPT_KEY to DELETE_OPERATION_OPT_VAL (DELETE_OPERATION_OPT_VAL has a string value of "delete"). In terms of method 3, I believe this flag method is only applicable when using the DeltaStreamer based on this blogpost https://hudi.apache.org/blog/delete-support-in-hudi/ , is this correct? And for method 1 of setting PAYLOAD_CLASS_OPT_KEY to "org.apache.hudi.EmptyHoodieRecordPayload", could you please clarify what the difference is between using method 1 and method 2? It seems the steps are the same as first a DataFrame must be prepared including the full records that need to be deleted, and then the write operation must be done using one of the 2 methods..

Reviewer: the flag method is not only applicable when using DeltaStreamer, but also applicable when using HoodieWriteClient/Spark DataSource, in this way, the ingested batch record would consist of both new insert records and delete records, that means you would insert and delete records with one batch, but when setting to `EmptyHoodieRecordPayload`, the ingested batch record is considered to be all delete records.

Author: Understood. Will clarify the Deletes section with this.

Available values:<br>
`UPSERT_OPERATION_OPT_VAL` (default), `BULK_INSERT_OPERATION_OPT_VAL`, `INSERT_OPERATION_OPT_VAL`, `DELETE_OPERATION_OPT_VAL`

**TABLE_TYPE_OPT_KEY**: The [type of table](/docs/concepts.html#table-types) to write to.<br>
Available values:<br>
[`COW_TABLE_TYPE_OPT_VAL`](/docs/concepts.html#copy-on-write-table) (default), [`MOR_TABLE_TYPE_OPT_VAL`](/docs/concepts.html#merge-on-read-table)

**KEYGENERATOR_CLASS_OPT_KEY**: Key generator class, that will extract the key out of incoming record. If single column key use `SimpleKeyGenerator`. For multiple column keys use `ComplexKeyGenerator`. Note: A custom key generator class can be written/provided here as well. Primary key columns should be provided via `RECORDKEY_FIELD_OPT_KEY` option.<br>
Available values:<br>

> `classOf[SimpleKeyGenerator].getName` (default), `classOf[NonpartitionedKeyGenerator].getName`, `classOf[ComplexKeyGenerator].getName`

Author: Add clarity on current limitation

```suggestion
`classOf[SimpleKeyGenerator].getName` (default), `classOf[NonpartitionedKeyGenerator].getName` (Non-partitioned tables can currently only have a single key column, [HUDI-1053](https://issues.apache.org/jira/browse/HUDI-1053)), `classOf[ComplexKeyGenerator].getName`
```

**HIVE_PARTITION_EXTRACTOR_CLASS_OPT_KEY**: Specify if the table should or should not be partitioned.<br>
Available values:<br>

> `classOf[MultiPartKeysValueExtractor].getCanonicalName` (default), `classOf[NonPartitionedExtractor].getCanonicalName`

Reviewer: `TimestampBasedKeyGenerator` and `GlobalDeleteKeyGenerator` are also available.

Author: Good call, also it seems that default is now actually `classOf[SlashEncodedDayPartitionValueExtractor].getCanonicalName`, which is also available.

```suggestion
`classOf[SlashEncodedDayPartitionValueExtractor].getCanonicalName` (default), `classOf[MultiPartKeysValueExtractor].getCanonicalName`, `classOf[TimestampBasedKeyGenerator].getCanonicalName`, `classOf[NonPartitionedExtractor].getCanonicalName`, `classOf[GlobalDeleteKeyGenerator].getCanonicalName` (to be used when OPERATION_OPT_KEY is set to DELETE_OPERATION_OPT_VAL)
```

> **HIVE_PARTITION_EXTRACTOR_CLASS_OPT_KEY**: Specify if the table should or should not be partitioned.<br>

Reviewer: same as this section.

Author: Also adding clarity that this is only necessary when using hive.

Example:
Upsert a DataFrame, specifying the necessary field names for `recordKey => _row_key`, `partitionPath => partition`, and `precombineKey => timestamp`

```java
inputDF.write()

>        .format("org.apache.hudi")

Reviewer: would be changed to `format("hudi")`

Author: This seems to be specified as "org.apache.hudi" by all the existing documentation that I can find and when I run the code as well. See example of current usage here: https://hudi.apache.org/docs/writing_data.html#datasource-writer

>        .format("org.apache.hudi")

Reviewer: would be changed to `format("hudi")`

Author: marking resolved as duplicate

 - **Soft Deletes** : With soft deletes, user wants to retain the key but just null out the values for all other fields. 
 This can be simply achieved by ensuring the appropriate fields are nullable in the table schema and simply upserting the table after setting these fields to null.
 
 - **Hard Deletes** : A stronger form of deletion is to physically remove any trace of the record from the table. This can be achieved in 3 different ways.

   1) Using DataSource, set `OPERATION_OPT_KEY` to `DELETE_OPERATION_OPT_VAL`. This will remove all records in the DataSet being submitted.
   
   2) Using DataSource, set `PAYLOAD_CLASS_OPT_KEY` to `"org.apache.hudi.EmptyHoodieRecordPayload"`. This will remove all records in the DataSet being submitted. 
   
   3) Using DataSource or DeltaStreamer, add a column named `_hoodie_is_deleted` to DataSet. The value of this column must be set to `true` for all records to be deleted and either `false` or left null for any records to be upserted.

>     

Reviewer: great.
